experimental/NeuralNets/Lib.idr

```diff
@@ -0,0 +1,151 @@
+
+export
+nums : String
+nums = "0123456789!@#$%^&*()_+=-:;"
+
+export
+removeStopWords : String -> List String
+removeStopWords line = do
+              let tmp = words line
+              foldr (\x, acc => if elem x stopwords then acc else x::acc ) [] tmp
+
+export
+makeWordPairs : List String -> List (String, String)
+makeWordPairs [] = []
+makeWordPairs (x::y::z::h::xs) = (x,y)::(x,z)::(y,z)::(y,h)::(z,x)::(z,y)::(z,h)::(h,x)::(h,y)::(h,z)::makeWordPairs xs
+makeWordPairs (x::y::z::xs) = (x,y)::(x,z)::(y,z)::(z,x)::(z,y)::makeWordPairs xs
+makeWordPairs (x::y::xs) = (x,y)::makeWordPairs xs
+
+export
+createDictionary : List String ->  List (String, Nat)
+createDictionary xss = func 0 tmp 
+              where
+                tmp = sort . nub $ xss
+                func : Nat -> List String -> List (String, Nat)
+                func _ [] = []
+                func Z (x::xs) = (x, S Z) :: func (S (S Z)) xs
+                func (S k) (x::xs) = (x, S k) :: func (S (S k) ) xs
+
+
+export
+insertAt : List Double -> Double -> Nat -> List Double
+insertAt [] elem pos = []
+insertAt (x::xs) elem  Z = elem::x::xs
+insertAt (x::xs) elem (S Z) = x::elem::xs
+insertAt (x::xs) elem (S k) = x::insertAt xs elem k 
+
+export
+oneHotEncoder : List Nat -> List Double -> List (List Double)
+oneHotEncoder ws zs = foldr(\x, acc => (insertAt zs 1.0 x)::acc) [] ws
+
+export
+matrix_gen : Num j =>  List j -> List j -> List (List j )
+matrix_gen ws zs = foldr (\x, acc => zs::acc) [] ws
+
+b : Double
+b = 0.3
+
+export
+softmax_activation : List Double -> List Double
+softmax_activation xs = let softmax_denom = sum $ map exp $ xs 
+in foldr (\x,acc => (exp x)/softmax_denom::acc) [] xs
+
+export
+nn_Layer : List Double -> List Double -> List Double
+nn_Layer [] _ = []
+nn_Layer _ [] = []
+nn_Layer [] [] = []
+nn_Layer (x::xs) (y::ys) = (x * y + b)::nn_Layer xs ys
+
+
+export
+grad_CrossEntropy : List Double -> List Double -> List Double
+grad_CrossEntropy [] _ = []
+grad_CrossEntropy _ [] = []
+grad_CrossEntropy [] [] = []
+grad_CrossEntropy (x::xs) (y::ys) = -(y/x) + (1-y)/(1-x)::grad_CrossEntropy xs ys
+
+export
+adam_Optimizer : List Double -> List Double -> Double -> Double -> Double -> Double -> Double -> Double -> List Double
+adam_Optimizer [] _ _ _ _ _ _ _ =  []
+adam_Optimizer _ [] _ _ _ _ _ _ =  []
+adam_Optimizer (t::thetas) (gt::gts) alpha m_prev v_prev beta1 beta2 epsilon = do
+              let m_current = (beta1 * m_prev + (1-beta1) * gt)/ (1-beta1)
+              let v_current = (beta2 * v_prev + (1-beta2) *  (db.pow gt 2))/(1-(db.pow beta2 2))
+              let mhat_current = m_current /(1-beta1)
+              let vhat_current = v_current/(1-beta2)
+              (t - alpha * mhat_current / ((sqrt  vhat_current) + epsilon)):: adam_Optimizer thetas gts alpha m_current v_current beta1 beta2 epsilon
+
+export
+wordpairs : List String -> List (String,String)
+wordpairs wrds = makeWordPairs wrds
```

Comment:
You don't need `wrds`, you can just write

```idris
wordpairs = makeWordPairs
```

Reply:
Correct!